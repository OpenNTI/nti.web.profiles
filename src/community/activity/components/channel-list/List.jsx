import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';
import {Text} from '@nti/web-commons';

import Styles from './Styles.css';
import Item from './Item';

const cx = classnames.bind(Styles);

ChannelList.propTypes = {
	list: PropTypes.shape({
		label: PropTypes.string,
		channels: PropTypes.array
	}).isRequired,
	activeChannel: PropTypes.object
};
export default function ChannelList ({list, activeChannel}) {
	const {label, channels} = list;

	return (
		<div className={cx('channel-list')}>
			{label && (
				<Text.Base as="div" className={cx('channel-header')}>
					{label}
				</Text.Base>
			)}
			<ul>
				{channels.map((channel) => {
					return (
						<li key={channel.getID()}>
							<Item channel={channel} active={activeChannel.getID() === channel.getID()} />
						</li>
					);
				})}
			</ul>
		</div>
	);
}