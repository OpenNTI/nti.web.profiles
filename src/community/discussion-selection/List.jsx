import React from 'react';
import PropTypes from 'prop-types';

import { filterItemsBySearchTerm } from './utils';
import { ListContainerBase, Empty } from './parts';
import NewTopic from './NewTopic';
import Topic from './Topic';
import Placeholder from './Placeholder';

const Container = styled(ListContainerBase)`
	margin-top: 30px;
`;


List.propTypes = {
	items: PropTypes.shape({
		hasMore: PropTypes.bool,
		more: PropTypes.func,
		[Symbol.iterator]: PropTypes.func,
	}),
	searchTerm: PropTypes.string,
	onCreate: PropTypes.func,
	onSelect: PropTypes.func,
};


export default function List ({ items, onCreate, onSelect, searchTerm, ...props}) {
	const filteredItems = filterItemsBySearchTerm(items, searchTerm);

	return (
		<Container data-testid="discussion-selection-list">

			{(!filteredItems?.length) && (
				<Empty data-testid="no-results">
					No discussions found
				</Empty>
			)}

			<NewTopic onClick={onCreate}/>

			{filteredItems.map((item, i) => (
				<Topic
					key={i}
					item={item}
					onClick={onSelect}
					searchTerm={searchTerm}
					{...props}
				/>
			)
			)}

			{items.hasMore && (
				<Placeholder key={filteredItems.length} load={items.more}/>
			)}
		</Container>
	);
}

