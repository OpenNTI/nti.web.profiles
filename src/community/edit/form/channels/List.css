.channels {
	list-style: none;
	margin: 0;
	padding: 0;
}

.channels li {
	padding: 0;
	margin: 0;
}