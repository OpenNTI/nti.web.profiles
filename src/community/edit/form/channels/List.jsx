import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';
import {Text} from '@nti/web-commons';

import Styles from './List.css';
import Channel from './Channel';

const cx = classnames.bind(Styles);

export default class ChannelsFieldList extends React.Component {
	static propTypes = {
		list: PropTypes.shape({
			label: PropTypes.string,
			channels: PropTypes.object,
			order: PropTypes.arrayOf(PropTypes.string)
		}),
		onChange: PropTypes.func,
		errors: PropTypes.object
	}

	state = {sortable: [], pinned: []}

	componentDidMount () {
		this.setup();
	}

	componentDidUpdate (prevProps) {
		const {list} = this.props;
		const {list:prevList} = prevProps;

		if (list !== prevList) {
			this.setup();
		}
	}


	setup () {
		const {list} = this.props;
		const {order, channels} = list;

		const {sortable, pinned} = order.reduce((acc, channelId) => {
			const channel = channels[channelId];

			if (channel.pinned) {
				acc.pinned.push(channel);
			} else {
				acc.sortable.push(channel);
			}

			return acc;
		}, {sortable: [], pinned: []});

		this.setState({
			sortable,
			pinned
		});
	}

	onChannelChange = (channel) => {
		const {list, onChange} = this.props;

		if (onChange) {
			onChange({
				...list,
				channels: {
					...list.channels,
					[channel.id]: channel
				}
			});
		}
	}

	render () {
		const {list} = this.props;
		const {sortable, pinned} = this.state;

		return (
			<div className={cx('channel-list')}>
				{list.label && (<Text.Base className={cx('label')}>{list.label}</Text.Base>)}
				{this.renderPinned(pinned)}
				{this.renderSortable(sortable)}
			</div>
		);
	}


	renderPinned (channels) {
		if (!channels || !channels.length) { return null; }

		return (
			<ol className={cx('channels')}>
				{channels.map((channel) => {
					return (
						<li key={channel.id}>
							{this.renderChannel(channel)}
						</li>
					);
				})}
			</ol>
		);
	}


	renderSortable (channels) {
		if (!channels || !channels.length) { return null; }

		//TODO: make this list sortable

		return (
			<ol className={cx('channels')}>
				{channels.map((channel) => {
					return (
						<li key={channel.id}>
							{this.renderChannel(channel)}
						</li>
					);
				})}
			</ol>
		);
	}


	renderChannel (channel) {
		const {errors} = this.props;

		return (
			<Channel channel={channel} onChange={this.onChannelChange} errors={errors && errors[channel.id]} />
		);
	}
}
