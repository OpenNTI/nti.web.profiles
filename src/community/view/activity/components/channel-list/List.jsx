import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames/bind';
import {Text} from '@nti/web-commons';

import Styles from './Styles.css';
import Item from './Item';

const cx = classnames.bind(Styles);

ChannelList.propTypes = {
	list: PropTypes.shape({
		label: PropTypes.string,
		channels: PropTypes.array
	}).isRequired,
	activeChannel: PropTypes.object
};
export default function ChannelList ({list, activeChannel}) {
	const {label, channels} = list;
	const activeId = activeChannel && activeChannel.getID();

	const {pinned, sortable} = channels.reduce((acc, channel) => {
		if (channel.pinned) {
			acc.pinned.push(channel);
		} else {
			acc.sortable.push(channel);
		}

		return acc;
	}, {pinned: [], sortable: []});
	
	const renderChannel = (channel) => {
		return (
			<li key={channel.getID()}>
				<Item channel={channel} active={activeId === channel.getID()} />
			</li>
		);
	};

	return (
		<div className={cx('channel-list', {'has-label': !!label})}>
			{label && (
				<Text.Base as="div" className={cx('channel-header')}>
					{label}
				</Text.Base>
			)}
			<ul>
				{pinned.map(renderChannel)}
				{sortable.map(renderChannel)}
			</ul>
		</div>
	);
}