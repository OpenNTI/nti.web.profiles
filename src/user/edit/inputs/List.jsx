import React from 'react';
import PropTypes from 'prop-types';
import {PropTypes as PT} from '@nti/lib-commons';
import {Button} from '@nti/web-commons';
import cx from 'classnames';
import {scoped} from '@nti/lib-locale';

const t = scoped('nti-web-profile.user-profile.edit.list', {
	add: 'Add Entry',
});

export default class List extends React.Component {

	static of = Component => (
		(props) => <List {...props} component={Component} />
	)

	static propTypes = {
		value: PropTypes.array,
		onChange: PropTypes.func,
		className: PropTypes.string,
		includeBlank: PropTypes.bool,
		component: PT.component
	}

	addEntry = () => {
		const {value, component, onChange} = this.props;
		const entry = component.createEmpty && component.createEmpty();
		onChange([...value, entry]);
	}

	removeEntry = (index) => {
		const {value: values, onChange} = this.props;
		values.splice(index, 1);
		onChange(values);
	}

	onChange = (value, index) => {
		const {value: values, onChange} = this.props;
		values[index] = value;
		onChange(values);
	}

	render () {
		const {
			className,
			component,
			includeBlank,
			value,
			...props
		} = this.props;

		const values = includeBlank
			? [...value, component.createEmpty && component.createEmpty()]
			: value;

		return (
			<div className={cx('nti-profile-field-list', className)}>
				<ul className="nti-profile-field-list-values">
					{(values || []).map((v, i) => (
						<li key={i}>
							<Item component={component} index={i} {...props} value={v} onChange={this.onChange} onRemove={this.removeEntry}/>
						</li>
					))}
				</ul>
				{!includeBlank && <Button className="add-entry" onClick={this.addEntry}>{t('add')}</Button>}
			</div>
		);
	}
}

class Item extends React.Component {

	static propTypes = {
		index: PropTypes.number.isRequired,
		onChange: PropTypes.func.isRequired,
		onRemove: PropTypes.func,
		component: PT.component
	}

	onChange = value => {
		const {onChange, index} = this.props;
		onChange(value, index);
	}

	onRemove = () => {
		const {onRemove, index} = this.props;
		onRemove(index);
	}

	render () {
		const {component: Component, onRemove, ...props} = this.props;
		return (
			<>
				{onRemove && <span className="remove icon-remove" onClick={this.onRemove} />}
				<Component {...props} onChange={this.onChange} />
			</>
		);
	}
}
