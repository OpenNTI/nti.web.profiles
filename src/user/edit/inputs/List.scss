@import "~@nti/style-common";

.nti-profile-field-list {
	.nti-profile-field-list-values {
		list-style: none;
		padding: 0;
		margin: 0;

		> * {
			position: relative;
			margin-bottom: 3rem;

			&:last-child {
				margin-bottom: 1rem;
			}

			.remove {
				position: absolute;
				top: 0;
				right: -20px;
				color: $tertiary-grey;
				cursor: pointer;

				&:hover + * {
					background-color: $quad-grey;
					box-shadow: 20px 0 0 6px $quad-grey, -10px 0 0 6px $quad-grey;
				}
			}
		}
	}
}
