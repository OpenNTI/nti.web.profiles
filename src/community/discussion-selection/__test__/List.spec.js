/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitFor, fireEvent } from '@testing-library/react';

import List from '../List';

describe('Topic List test', () => {

	test('Filtered', async () => {
		const items = [
			{ title: 'item 1' },
			{ title: 'item 2' },
			{ title: 'item 3' }
		];

		const {getAllByTestId} = render(<List items={items} searchTerm="2"/>);
		expect(getAllByTestId('discussion-selection-topic-title').length).toBe(1);

		const tree = renderer.create(<List items={items} searchTerm="2" />).toJSON();
		expect(tree).toMatchSnapshot();
	});


	test('Simple list', async () => {
		const student = {
			alias: 'student'
		};

		const instructor = {
			alias: 'instructor'
		};

		const topics = [
			{ title: 'item 1', user: student },
			{ title: 'item 2', user: student },
			{ title: 'item 3', user: instructor }
		];

		const selectedTopics = new Set();

		const onTopicSelect = jest.fn();

		const results = render(<List items={topics} selected={selectedTopics} onSelect={onTopicSelect}/>);
		expect(results.getAllByText(/item \d/).length).toBe(3);

		fireEvent.click(await results.queryByText(/item 1/));

		await waitFor(() =>
			expect(onTopicSelect).toHaveBeenCalled());
	});
});
