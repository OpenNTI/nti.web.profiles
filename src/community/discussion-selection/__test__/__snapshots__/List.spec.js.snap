// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic List test Filtered 1`] = `
<div
  data-testid="discussion-selection-list"
  layout="grid"
>
  <div
    data-testid="discussion-selection-item"
    layout="grid"
  >
    <div
      layout="grid"
    >
      <i
        className="icon-discuss"
        layout="grid"
      />
    </div>
    <span
      className="ntiText"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Start a new discussion
in this channel",
        }
      }
    />
    <span
      layout="grid"
    />
    <svg
      className="svg-icon"
      fill="currentColor"
      height="26"
      layout="grid"
      viewBox="0 0 26 26"
      width="26"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fillRule="evenodd"
        transform="translate(1 1)"
      >
        <circle
          cx="12"
          cy="12"
          r="12"
          stroke="currentColor"
        />
        <path
          d="M13 5v6h6v2h-6v6h-2v-6H5v-2h6V5h2z"
          fill="currentColor"
        />
      </g>
    </svg>
  </div>
  <div
    data-testid="discussion-selection-item"
    layout="grid"
    onClick={[Function]}
  >
    <div
      data-testid="discussion-selection-avatar"
      layout="grid"
      style={
        Object {
          "backgroundImage": "url(\\"/app/resources/images/elements/discussion-icon.png\\")",
        }
      }
    />
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<span>item </span><span class=\\"highlight\\">2</span>",
        }
      }
      data-testid="discussion-selection-topic-title"
      layout="grid"
    />
    <span
      layout="grid"
    />
    <time
      className="ntiText"
      dateTime="2021-01-28T01:43:48Z"
    >
      a few seconds ago
    </time>
  </div>
</div>
`;
